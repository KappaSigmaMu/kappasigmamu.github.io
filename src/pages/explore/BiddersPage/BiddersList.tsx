import Identicon from '@polkadot/react-identicon'
import type { Vec } from '@polkadot/types'
import type { PalletSocietyBid } from '@polkadot/types/lookup'
import { useState } from 'react'
import { Col, Badge } from 'react-bootstrap'
import styled from 'styled-components'
import { DataHeaderRow, DataRow } from '../../../components/base'
import { FormatBalance } from '../../../components/FormatBalance'
import { humanizeBidKind } from '../../../helpers/humanize'
import { truncateMiddle } from '../../../helpers/truncate'
import { useKusama } from '../../../kusama'
import { ApiState } from '../../../kusama/KusamaContext'
import { unbid, unvouch } from './helper'

type Props = { bids: Vec<PalletSocietyBid>, activeAccount: accountType; handleResult: any }
type OnStatusChangeProps = { loading : boolean, message : string, success: boolean } 

// TODO: move this to a `components` directory to follow the convention of other pages 
const BiddersList = ({ bids, activeAccount, handleResult } : Props) : JSX.Element => {
  const { api, apiState } = useKusama()
  const [loading, setLoading] = useState(false)

  const isBidder = (bid : PalletSocietyBid) => activeAccount?.address === bid.who.toString()
  const isVoucher = (bid : PalletSocietyBid) => activeAccount?.address === bid.kind.asVouch?.[0].toString()

  const apiReady = apiState === ApiState.ready

  const onStatusChange = ({ loading, message, success } : OnStatusChangeProps) => {
    setLoading(loading)
    handleResult({ message, success })
  }

  const handleUnbid = (index : any) => {
    const tx = api?.tx?.society?.unbid(index)
    apiReady && unbid(tx, activeAccount, onStatusChange)
  }

  const handleUnvouch = (index : any) => {
    const tx = api?.tx?.society?.unvouch(index)
    apiReady && unvouch(tx, activeAccount, onStatusChange)
  }

  const ownerActions = (bid : PalletSocietyBid) => {
    let pillText, handleUndo : any, badgeText

    if (bid.kind.isDeposit && isBidder(bid)) {
      pillText = 'My bid'
      handleUndo = handleUnbid
      badgeText = 'UNBID'
    } else if (bid.kind.isVouch && isVoucher(bid)) {
      pillText = 'My vouch'
      handleUndo = handleUnvouch
      badgeText = 'UNVOUCH'
    }

    return { pillText, handleUndo, badgeText }
  }

  const BidVouchIdentifier = ({ bid, index } : { bid: PalletSocietyBid, index : number }) => {
    const { pillText, badgeText, handleUndo } = ownerActions(bid)

    return (
      <>
        <Col xs={2}>
          {<FormatBalance balance={bid.value} />}
        </Col>
        <Col xs={2}>
          {bid.kind.isVouch && <FormatBalance balance={bid.kind.asVouch?.[1]} />}
        </Col>
        <Col xs={2}>
          {badgeText &&
            <>
              <StyledUndo disabled={loading} onClick={() => handleUndo(index)} href="#">
                {badgeText}
              </StyledUndo>
              <Badge pill bg="primary">
                {pillText}
              </Badge>
            </>
          }
        </Col>
      </>
    )
  }

  const isOwner = (bid : PalletSocietyBid) => {
    if (bid.kind.isDeposit) {
      return isBidder(bid)
    } else if (bid.kind.isVouch) {
      return isVoucher(bid)
    }
  }

  if (bids.length === 0) return <>No bids</>

  return (
    <>
      <DataHeaderRow>
        <Col xs={1} className="text-center">#</Col>
        <Col xs={3} className="text-start">Wallet Hash</Col>
        <Col xs={2} className="text-start">Bid Kind</Col>
        <Col xs={2} className="text-start">Value</Col>
        <Col xs={4} className="text-start">Tip</Col>
      </DataHeaderRow>
      {bids.map((bid : PalletSocietyBid, index : any) => (
        <StyledDataRow isOwner={isOwner(bid)} key={bid.who?.toString()}>
          <Col xs={1} className="text-center">
            <Identicon value={bid.who} size={32} theme={'polkadot'} />
          </Col>
          <Col xs={3} className="text-start text-truncate">
            {truncateMiddle(bid.who?.toString())}
          </Col>
          <Col xs={2} className="text-start text-truncate">
            {humanizeBidKind(bid.kind)}
          </Col>
          <BidVouchIdentifier bid={bid} index={index} />
        </StyledDataRow>
      ))}
    </>
  )
}

const StyledDataRow = styled(DataRow)`
  background-color: ${(props) => props.isOwner ? '#73003d' : ''};
  border: ${(props) => props.isOwner ? '2px solid #E6007A' : ''};
`

type PropsUnbid = {
  disabled: boolean;
};

const StyledUndo = styled.a.attrs((props : PropsUnbid) => ({
  disabled: props.disabled
}))<PropsUnbid>`
  color: ${(props) => props.disabled ? 'grey' : '#E6007A'};
  margin-right: 3%;
  font-weight: 800;
  font-size: 13px;
  pointer-events: ${(props) => props.disabled ? 'none' : ''};
`

export { BiddersList }
