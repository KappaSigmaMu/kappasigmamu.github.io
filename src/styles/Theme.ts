import { DefaultTheme } from 'styled-components'

const Theme: DefaultTheme = {
  darkBg: '#212529',
  lightBg: '#ffffff',

  colors: {
    primary: '#E6007A',
    secondary: '#01FFFF',
    white: '#FFFFFF',
    black: '#212529',
    grey: '#ced4da',
    lightGrey: '#6c757d',
    darkGrey: '#343A40',
  },
}

export { Theme }
