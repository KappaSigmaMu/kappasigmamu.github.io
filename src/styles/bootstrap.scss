$theme-colors: (
  'primary': #e6007a,
  'secondary': #01ffff,
  // 'success': red,
  // 'info': red,
  // 'warning': red,
  // 'danger': rgb(204, 193, 193),
  'light': white,
  'dark': black,
  'grey-dark': #6c757d,
);

@import '~bootstrap/scss/bootstrap.scss';
