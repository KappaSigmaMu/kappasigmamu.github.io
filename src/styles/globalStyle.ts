import { createGlobalStyle } from 'styled-components'

const GlobalStyle = createGlobalStyle`
  * {
    color: white;
  }

  html {
    background-color: #212529;
    font-family: Roboto, sans-serif !important;
    font-style: normal;
  }

  body {
    background-color: #212529;
    overflow: auto;
  }

  h1 {
    font-size: 36px;
    font-weight: bold;
  }

  h4 {
    font-size: 16px;
    font-weight: bold;
  }

  h2, h3, h4, h5 {
    color: white;
  }

  a {
    color: #01ffff;
    text-decoration: none;
  }

  a:hover {
    color: white;
  }

  .navbar-light .navbar-nav .nav-link {
    color: #01ffff;
  }

  p {
    color: #c4c4c4;
  }
`

export { GlobalStyle }
