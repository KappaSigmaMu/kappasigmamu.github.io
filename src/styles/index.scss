// Override default variables before the import
$theme-colors: (
  'primary': #e6007a,
  'secondary': #01ffff,
  'dark': black,
  'grey-dark': #6c757d,
  'light': white,
);

* {
  color: white;
}

.dropdown-toggle::after {
  vertical-align: 36% !important;
  margin-left: 0.7em !important;
}

html * {
  font-family: Roboto, sans-serif !important;
  font-style: normal !important;
}

h1 {
  font-weight: bold !important;
  font-size: 36px !important;
}

h4 {
  font-weight: bold !important;
  font-size: 16px !important;
}

h2, h3, h4, h5 {
  color: white;
}

p {
  color: #C4C4C4;
}

a {
  color: #01ffff !important;
}

.btn-lg {
  font-weight: bold !important;
}

.btn-link {
  color: #01ffff !important;
}

// Import Bootstrap and its default variables
@import '~bootstrap/scss/bootstrap.scss';
