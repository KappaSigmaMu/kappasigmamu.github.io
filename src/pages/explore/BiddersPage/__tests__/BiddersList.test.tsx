import TestRenderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import { PalletSocietyBidFixture } from '../../../../__tests__/fixtures/PalletSocietyBidFixture'
import { Theme } from '../../../../styles/Theme'
import { BiddersList } from '../BiddersList'

const bidDeposit = PalletSocietyBidFixture.createDeposit('5GrpknVvGGrGH3EFuURXeMrWHvbpj3VfER1oX5jFtuGbfzCE', 20)
const bidVouch = PalletSocietyBidFixture.createDeposit('CxDDSH8gS7jecsxaRL9Txf8H5kqesLXAEAEgp76Yz632J9M', 10)
const bids = [bidVouch, bidDeposit]

test('renders BiddersList', () => {
  const tree = TestRenderer.create(
    <ThemeProvider theme={Theme}>
      {/* <BiddersList bids={bids} /> */}
    </ThemeProvider>
  )

  expect(tree).toMatchSnapshot()
})
