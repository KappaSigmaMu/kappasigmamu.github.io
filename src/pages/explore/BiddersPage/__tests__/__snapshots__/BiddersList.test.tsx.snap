// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders BiddersList 1`] = `
Array [
  <div
    className="sc-bdfBQB DMQmf row"
  >
    <div
      className="text-center col-1"
    >
      #
    </div>
    <div
      className="text-start col-4"
    >
      Wallet Hash
    </div>
    <div
      className="text-start col-2"
    >
      Bid Kind
    </div>
    <div
      className="text-end col-4"
      style={
        Object {
          "paddingRight": 0,
        }
      }
    >
      Value
    </div>
  </div>,
  <div
    className="sc-bdfBQB sc-gsTEea sc-jSgvzq DMQmf frwqlw gQQaIy row"
    isBidder={false}
  >
    <div
      className="text-center col-1"
    >
      <div
        className="Identicon__Wrapper-sc-1gm2vek-0 iULmUS ui--IdentityIcon  "
        onClick={[Function]}
      >
        <svg
          className=""
          height={32}
          id="5CSbZ7wG456oty4WoiX6a1J88VUbrCXLhrKVJ9q95BsYH4TZ"
          name="5CSbZ7wG456oty4WoiX6a1J88VUbrCXLhrKVJ9q95BsYH4TZ"
          viewBox="0 0 64 64"
          width={32}
        >
          <circle
            cx={32}
            cy={32}
            fill="#eee"
            r={32}
          />
          <circle
            cx={32}
            cy={8}
            fill="hsl(90, 94%, 35%)"
            r={5}
          />
          <circle
            cx={32}
            cy={20}
            fill="hsl(253, 94%, 15%)"
            r={5}
          />
          <circle
            cx={21.607695154586736}
            cy={14}
            fill="hsl(286, 94%, 53%)"
            r={5}
          />
          <circle
            cx={11.215390309173472}
            cy={20}
            fill="hsl(84, 94%, 15%)"
            r={5}
          />
          <circle
            cx={21.607695154586736}
            cy={26}
            fill="hsl(275, 94%, 15%)"
            r={5}
          />
          <circle
            cx={11.215390309173472}
            cy={32}
            fill="hsl(286, 94%, 53%)"
            r={5}
          />
          <circle
            cx={11.215390309173472}
            cy={44}
            fill="hsl(90, 94%, 35%)"
            r={5}
          />
          <circle
            cx={21.607695154586736}
            cy={38}
            fill="hsl(253, 94%, 15%)"
            r={5}
          />
          <circle
            cx={21.607695154586736}
            cy={50}
            fill="hsl(292, 94%, 15%)"
            r={5}
          />
          <circle
            cx={32}
            cy={56}
            fill="hsl(163, 94%, 35%)"
            r={5}
          />
          <circle
            cx={32}
            cy={44}
            fill="hsl(61, 94%, 53%)"
            r={5}
          />
          <circle
            cx={42.392304845413264}
            cy={50}
            fill="hsl(225, 94%, 35%)"
            r={5}
          />
          <circle
            cx={52.78460969082653}
            cy={44}
            fill="hsl(253, 94%, 15%)"
            r={5}
          />
          <circle
            cx={42.392304845413264}
            cy={38}
            fill="hsl(309, 94%, 75%)"
            r={5}
          />
          <circle
            cx={52.78460969082653}
            cy={32}
            fill="hsl(225, 94%, 35%)"
            r={5}
          />
          <circle
            cx={52.78460969082653}
            cy={20}
            fill="hsl(163, 94%, 35%)"
            r={5}
          />
          <circle
            cx={42.392304845413264}
            cy={26}
            fill="hsl(61, 94%, 53%)"
            r={5}
          />
          <circle
            cx={42.392304845413264}
            cy={14}
            fill="hsl(292, 94%, 15%)"
            r={5}
          />
          <circle
            cx={32}
            cy={32}
            fill="hsl(298, 94%, 15%)"
            r={5}
          />
        </svg>
      </div>
    </div>
    <div
      className="text-start text-truncate col-4"
    >
      5CSbZ7...sYH4TZ
    </div>
    <div
      className="text-start text-truncate col-2"
    >
      Deposit
    </div>
    <div
      className="text-end col-4"
      style={
        Object {
          "paddingRight": 0,
        }
      }
    >
       
      <span>
        10.
      </span>
      <span>
        0000
      </span>
      <span>
         
        KSM
      </span>
    </div>
    <div
      className="text-start col-1"
    />
  </div>,
  <div
    className="sc-bdfBQB sc-gsTEea sc-jSgvzq DMQmf frwqlw gQQaIy row"
    isBidder={false}
  >
    <div
      className="text-center col-1"
    >
      <div
        className="Identicon__Wrapper-sc-1gm2vek-0 iULmUS ui--IdentityIcon  "
        onClick={[Function]}
      >
        <svg
          className=""
          height={32}
          id="5GrpknVvGGrGH3EFuURXeMrWHvbpj3VfER1oX5jFtuGbfzCE"
          name="5GrpknVvGGrGH3EFuURXeMrWHvbpj3VfER1oX5jFtuGbfzCE"
          viewBox="0 0 64 64"
          width={32}
        >
          <circle
            cx={32}
            cy={32}
            fill="#eee"
            r={32}
          />
          <circle
            cx={32}
            cy={8}
            fill="hsl(118, 36%, 15%)"
            r={5}
          />
          <circle
            cx={32}
            cy={20}
            fill="hsl(258, 36%, 15%)"
            r={5}
          />
          <circle
            cx={21.607695154586736}
            cy={14}
            fill="hsl(135, 36%, 75%)"
            r={5}
          />
          <circle
            cx={11.215390309173472}
            cy={20}
            fill="hsl(354, 36%, 35%)"
            r={5}
          />
          <circle
            cx={21.607695154586736}
            cy={26}
            fill="hsl(140, 36%, 75%)"
            r={5}
          />
          <circle
            cx={11.215390309173472}
            cy={32}
            fill="hsl(258, 36%, 15%)"
            r={5}
          />
          <circle
            cx={11.215390309173472}
            cy={44}
            fill="hsl(354, 36%, 35%)"
            r={5}
          />
          <circle
            cx={21.607695154586736}
            cy={38}
            fill="hsl(140, 36%, 75%)"
            r={5}
          />
          <circle
            cx={21.607695154586736}
            cy={50}
            fill="hsl(135, 36%, 75%)"
            r={5}
          />
          <circle
            cx={32}
            cy={56}
            fill="hsl(118, 36%, 15%)"
            r={5}
          />
          <circle
            cx={32}
            cy={44}
            fill="hsl(258, 36%, 15%)"
            r={5}
          />
          <circle
            cx={42.392304845413264}
            cy={50}
            fill="hsl(22, 36%, 15%)"
            r={5}
          />
          <circle
            cx={52.78460969082653}
            cy={44}
            fill="hsl(208, 36%, 75%)"
            r={5}
          />
          <circle
            cx={42.392304845413264}
            cy={38}
            fill="hsl(286, 36%, 15%)"
            r={5}
          />
          <circle
            cx={52.78460969082653}
            cy={32}
            fill="hsl(337, 36%, 75%)"
            r={5}
          />
          <circle
            cx={52.78460969082653}
            cy={20}
            fill="hsl(208, 36%, 75%)"
            r={5}
          />
          <circle
            cx={42.392304845413264}
            cy={26}
            fill="hsl(286, 36%, 15%)"
            r={5}
          />
          <circle
            cx={42.392304845413264}
            cy={14}
            fill="hsl(22, 36%, 15%)"
            r={5}
          />
          <circle
            cx={32}
            cy={32}
            fill="hsl(16, 36%, 75%)"
            r={5}
          />
        </svg>
      </div>
    </div>
    <div
      className="text-start text-truncate col-4"
    >
      5Grpkn...GbfzCE
    </div>
    <div
      className="text-start text-truncate col-2"
    >
      Deposit
    </div>
    <div
      className="text-end col-4"
      style={
        Object {
          "paddingRight": 0,
        }
      }
    >
       
      <span>
        20.
      </span>
      <span>
        0000
      </span>
      <span>
         
        KSM
      </span>
    </div>
    <div
      className="text-start col-1"
    />
  </div>,
]
`;
